### PR TITLE
Document ccr-stats telemetry device

### PR DESCRIPTION
**Pull request by dliappis**
Also use fields prefixed by `shard.` for the values returned by
`_ccr/stats` and move "name:ccr-stats" to be at the top level of the
document (used to be "meta.name:ccr-stats").

### CODE REVIEW

File: esrally/mechanic/telemetry.py [hunk 1]
```diff
@@ -431,14 +431,17 @@ def record_stats_per_index(self, name, stats):
 
         for shard_stats in stats:
             if "shard_id" in shard_stats:
+                doc = {
+                    "name": "ccr-stats",
+                    "shard": shard_stats
+                }
                 shard_metadata = {
                     "cluster": self.cluster_name,
                     "index": name,
```
Comment (danielmitterdorfer) [inline esrally/mechanic/telemetry.py]: This seems to be included in the shard stats already. Does it make sense to duplicate this information?

File: esrally/mechanic/telemetry.py [hunk 2]
```diff
@@ -431,14 +431,17 @@ def record_stats_per_index(self, name, stats):
...
-                    "shard": shard_stats["shard_id"],
-                    "name": "ccr-stats"
+                    "shard": shard_stats["shard_id"]
```
Comment (danielmitterdorfer) [inline esrally/mechanic/telemetry.py]: This is included in the shard stats already. Does it make sense to duplicate this information?

File: esrally/mechanic/telemetry.py (hunk 1 above)
Comment (dliappis) [reply esrally/mechanic/telemetry.py]: The shard metrics, as reported by the CCR stats endpoint are associated with each following index and I wanted to retain the structure in this respect.

For example in the following excerpt, we can see this duplication, a key `index` and then per shard the more specific `leader-index` / `follower_index` keys. I feel that for the sake of maintaining this structure it makes sense to keep this duplication.

```
   "indices" : [
      {
        "index" : "geonames-copy",
        "shards" : [
          {
            "remote_cluster" : "leader",
            "leader_index" : "geonames",
            "follower_index" : "geonames-copy",
            "shard_id" : 0,
            "leader_global_checkpoint" : -1,
            "leader_max_seq_no" : -1,
            "follower_global_checkpoint" : -1,
            "follower_max_seq_no" : -1,
            "last_requested_seq_no" : -1,
            "outstanding_read_requests" : 1,
            "outstanding_write_requests" : 0,
            "write_buffer_operation_count" : 0,
            "write_buffer_size_in_bytes" : 0,
            "follower_mapping_version" : 1,
            "follower_settings_version" : 1,
            "total_read_time_millis" : 0,
            "total_read_remote_exec_time_millis" : 0,
            "successful_read_requests" : 0,
            "failed_read_requests" : 0,
            "operations_read" : 0,
            "bytes_read" : 0,
            "total_write_time_millis" : 0,
            "successful_write_requests" : 0,
            "failed_write_requests" : 0,
            "operations_written" : 0,
            "read_exceptions" : [ ],
            "time_since_last_read_millis" : 4004
          },
          {
            "remote_cluster" : "leader",
            "leader_index" : "geonames",
            "follower_index" : "geonames-copy",
            "shard_id" : 1,
            "leader_global_checkpoint" : -1,
            "leader_max_seq_no" : -1,
            "follower_global_checkpoint" : -1,
            "follower_max_seq_no" : -1,
            "last_requested_seq_no" : -1,
            "outstanding_read_requests" : 1,
            "outstanding_write_requests" : 0,
            "write_buffer_operation_count" : 0,
            "write_buffer_size_in_bytes" : 0,
            "follower_mapping_version" : 1,
            "follower_settings_version" : 1,
            "total_read_time_millis" : 0,
            "total_read_remote_exec_time_millis" : 0,
            "successful_read_requests" : 0,
            "failed_read_requests" : 0,
            "operations_read" : 0,
            "bytes_read" : 0,
            "total_write_time_millis" : 0,
            "successful_write_requests" : 0,
            "failed_write_requests" : 0,
            "operations_written" : 0,
            "read_exceptions" : [ ],
            "time_since_last_read_millis" : 4087
          },
```

File: esrally/mechanic/telemetry.py (hunk 2 above)
Comment (dliappis) [reply esrally/mechanic/telemetry.py]: Whereas here I think it doesn't make sense to duplicate the shard_id; I'll remove.

File: esrally/mechanic/telemetry.py (hunk 1 above)
Comment (danielmitterdorfer) [reply esrally/mechanic/telemetry.py]: Sure, I'm fine with that.

File: esrally/mechanic/telemetry.py [hunk 3]
```diff
@@ -31,7 +31,8 @@
 
 def list_telemetry():
     console.println("Available telemetry devices:\n")
-    devices = [[device.command, device.human_name, device.help] for device in [JitCompiler, Gc, FlightRecorder, PerfStat, NodeStats, RecoveryStats]]
+    devices = [[device.command, device.human_name, device.help] for device in [JitCompiler, Gc, FlightRecorder, PerfStat,
```
Comment (danielmitterdorfer) [inline esrally/mechanic/telemetry.py]: Should we also update the [docs](https://esrally.readthedocs.io/en/stable/telemetry.html) where we show an example of the listed telemetry devices?

**Comment by dliappis** [comment]
@danielmitterdorfer Could you please take another pass esp now that I've added the ccr recipe in https://github.com/elastic/rally/pull/676/commits/e592f1775de60b5f967f7480eac341d4c4b18757 and the corresponding documentation in https://github.com/elastic/rally/pull/676/commits/5ca587aa0ff1b7d012a5490dbb37605482841d6c?

File: docs/recipes.rst [hunk 4]
```diff
@@ -148,3 +148,29 @@ Also don't forget to rename the folder of your local working copy as Rally will
     mv teamtrackrepo default
 
 From now on, Rally will treat your repository as default and you need to run Rally with ``--track-repository=old-rally-default`` if you want to use the out-of-the-box Rally tracks.
+
+
+.. _recipe_testing_rally_against_ccr_clusters:
+
+Testing Rally against CCR Clusters using a remote metric store
```
Comment (danielmitterdorfer) [inline docs/recipes.rst]: Nit: clusters (lower case)

File: docs/recipes.rst [hunk 5]
```diff
@@ -148,3 +148,29 @@ Also don't forget to rename the folder of your local working copy as Rally will
...
+
+Testing Rally features (such as the ``ccr-stats`` telemetry device) requiring Elasticsearch clusters configured for `cross-cluster replication <https://www.elastic.co/guide/en/elastic-stack-overview/current/ccr-getting-started.html>`_ can be a time consuming process.
+
+For this reason we've created an example under `recipes/ccr in Rally's repository <https://github.com/elastic/rally/tree/master/recipes/ccr>`_ simplifying it.
+
+Running the ``start.sh`` script requires Docker locally installed and performs the following actions:
+
+1. Starts a small single node Elasticsearch cluster locally, to serve as a :ref:`metrics store <configuration_options>`. It also starts Kibana attached to the Elasticsearch metric store cluster.
+2. Creates a new configuration file for Rally under ``~/.rally/rally-metricstore.ini`` referencing Elasticsearch from step 1.
+3. Starts two additional small local Elasticsearch clusters with 1 node each, (version ``7.0.0`` by default) called ``leader`` and ``follower`` listening at ports 329201 and 32902 respectively.
```
Comment (danielmitterdorfer) [inline docs/recipes.rst]: `329201` -> `39201`

File: docs/recipes.rst [hunk 6]
```diff
@@ -148,3 +148,29 @@ Also don't forget to rename the folder of your local working copy as Rally will
...
+4. Accepts the trial license.
+5. Configures ``leader`` on the ``follower`` as a `remote cluster <https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-remote-clusters.html#configuring-remote-clusters>`_.
+6. Sets an `auto-follow pattern <https://www.elastic.co/guide/en/elasticsearch/reference/current/ccr-put-auto-follow-pattern.html#ccr-put-auto-follow-pattern>`_ on the follower for every index on the leader to be replicated as ``<leader-index-name>-copy``.
+7. Runs the `geonames track <https://github.com/elastic/rally-tracks/tree/master/geonames>`_, `append-no-conflicts-index-only challenge <https://github.com/elastic/rally-tracks/blob/d4814aa7bf54a9dafd4c77be076d54500c3f2dd4/geonames/challenges/default.json#L188-L222>`_ challenge, ingesting only 20% of the corpus. It also enables the ``ccr-stats`` :doc:`telemetry device </telemetry>` with a sample rate interval of ``1s``.
+
+Rally will push metrics to the metric store configured in 1. and they can be visualized by accessing Kibana at `http://locahost:5601 <http://localhost:5601>`_.
+
+To tear down everything just issue ``./stop.sh``.
```
Comment (danielmitterdorfer) [inline docs/recipes.rst]: Remove "just"

File: docs/recipes.rst [hunk 7]
```diff
@@ -148,3 +148,29 @@ Also don't forget to rename the folder of your local working copy as Rally will
...
+
+
+.. _recipe_testing_rally_against_ccr_clusters:
+
+Testing Rally against CCR Clusters using a remote metric store
+--------------------------------------------------------------
+
+Testing Rally features (such as the ``ccr-stats`` telemetry device) requiring Elasticsearch clusters configured for `cross-cluster replication <https://www.elastic.co/guide/en/elastic-stack-overview/current/ccr-getting-started.html>`_ can be a time consuming process.
+
+For this reason we've created an example under `recipes/ccr in Rally's repository <https://github.com/elastic/rally/tree/master/recipes/ccr>`_ simplifying it.
```
Comment (danielmitterdorfer) [inline docs/recipes.rst]: How about something along these lines?

```
Use `recipes/ccr in Rally's repository <https://github.com/elastic/rally/tree/master/recipes/ccr>`_ to test a simple but complete example.
```

(I'd also move this to the previous paragraph)

File: docs/recipes.rst [hunk 8]
```diff
@@ -148,3 +148,29 @@ Also don't forget to rename the folder of your local working copy as Rally will
...
+
+Running the ``start.sh`` script requires Docker locally installed and performs the following actions:
+
+1. Starts a small single node Elasticsearch cluster locally, to serve as a :ref:`metrics store <configuration_options>`. It also starts Kibana attached to the Elasticsearch metric store cluster.
```
Comment (danielmitterdorfer) [inline docs/recipes.rst]: Should we maybe state what "small" means? (i.e. 512MB heap for the metrics store and 1 GB heap for the regular nodes?)

File: recipes/ccr/start.sh [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+source .elastic-version
+
+# Start metrics store
+docker-compose -f ./metricstore-docker-compose.yml up -d 
+
+# Start Elasticsearch
+docker-compose up -d
+
+printf "Waiting for clusters to get ready "
```
Comment (danielmitterdorfer) [inline recipes/ccr/start.sh]: Is there a reason you're sometimes using `printf` and sometimes `echo`?

Comment (dliappis) [reply recipes/ccr/start.sh]: Yes to avoid newline characters printed (to facilitate printing `...` as progress)

File: docs/recipes.rst (hunk 7 above)
Comment (dliappis) [reply docs/recipes.rst]: So no need to provide links to cross-cluster-replication from the Elastic website i.e. get rid of the entire line 158 completely, right?

Comment (danielmitterdorfer) [reply docs/recipes.rst]: No. I'd just turn this into a single paragraph. I don't think that we should have two paragraphs here. So line 158 would stay.

File: docs/recipes.rst (hunk 8 above)
Comment (dliappis) [reply docs/recipes.rst]: Fixed in https://github.com/elastic/rally/pull/676/commits/7202a0019a53e0220015b8928ef981d823d38241

File: docs/recipes.rst (hunk 7 above)
Comment (dliappis) [reply docs/recipes.rst]: Thanks, fixed in https://github.com/elastic/rally/pull/676/commits/7202a0019a53e0220015b8928ef981d823d38241

File: docs/recipes.rst (hunk 5 above)
Comment (dliappis) [reply docs/recipes.rst]: Fixed in https://github.com/elastic/rally/pull/676/commits/7202a0019a53e0220015b8928ef981d823d38241

File: docs/recipes.rst (hunk 4 above)
Comment (dliappis) [reply docs/recipes.rst]: Fixed in https://github.com/elastic/rally/pull/676/commits/7202a0019a53e0220015b8928ef981d823d38241

**Pull request merged**